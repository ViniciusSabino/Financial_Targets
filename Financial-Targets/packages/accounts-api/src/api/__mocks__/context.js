export default {
    request: {
        method: "POST",
        url: "/accounts",
        header: {
            "content-type": "application/json",
            "user-agent": "PostmanRuntime/7.11.0",
            accept: "*/*",
            "cache-control": "no-cache",
            "postman-token": "1a806c31-c425-4d9f-9fec-c63607620960",
            host: "localhost:8080",
            "accept-encoding": "gzip, deflate",
            "content-length": "222",
            connection: "keep-alive",
        },
    },
    response: {
        status: 200,
        message: "Ok",
        header: {},
    },
    app: {
        subdomainOffset: 2,
        proxy: false,
        env: "development",
    },
    originalUrl: "/accounts",
    req: "<original node req>",
    res: "<original node res>",
    socket: "<original node socket>",
};
