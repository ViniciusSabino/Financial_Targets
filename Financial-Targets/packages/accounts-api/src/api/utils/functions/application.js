const buildTheResult = (data, page = 1, perPage = 20) => ({
    count: data?.length ? data.length : 0,
    page,
    perPage,
    data,
});

export { buildTheResult };
