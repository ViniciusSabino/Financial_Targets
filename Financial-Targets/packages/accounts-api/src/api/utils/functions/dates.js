import moment from "moment";
import { applicationEnum } from "../enumerators";

const { differences } = applicationEnum.date;

const getCurrentDate = () => moment().format();

const getDaysInCurrentMonth = () => moment().daysInMonth();

const createMomentDate = (date) => moment(date);

const getDateDifference = (differenceType, dateA, dateB) => {
    switch (differenceType) {
        case differences.milliseconds:
            return dateA.diff(dateB);
        case differences.minutes:
            return dateA.diff(dateB, differences.minutes);
        case differences.hours:
            return dateA.diff(dateB, differences.hours);
        case differences.days:
            return dateA.diff(dateB, differences.days);
        case differences.weeks:
            return dateA.diff(dateB, differences.weeks);
        case differences.months:
            return dateA.diff(dateB, differences.months);
        case differences.years:
            return dateA.diff(dateA, differences.years, true);
        default:
    }
};

const remainingPeriod = (date) => {
    const currentDate = getCurrentDate();
    return {
        inSeconds: date.diff(currentDate),
        inMinutes: date.diff(currentDate, differences.minutes),
        inHours: date.diff(currentDate, differences.hours),
        onDays: date.diff(currentDate, differences.days),
        inWeeks: date.diff(currentDate, differences.weeks),
        inMonths: date.diff(currentDate, differences.months),
        inYears: date.diff(currentDate, differences.years, true),
    };
};

export default {
    getCurrentDate,
    getDaysInCurrentMonth,
    createMomentDate,
    getDateDifference,
    remainingPeriod,
};
