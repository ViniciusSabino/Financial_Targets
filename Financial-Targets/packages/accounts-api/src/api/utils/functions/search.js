import { application } from "../enumerators";

const createFilterConditions = (params, allFilters) => {
    const keys = Object.keys(params);
    const values = Object.values(params);
    const conditions = {};

    keys.forEach((key, index) => {
        const filter = allFilters.find((f) => f.parameter === key);
        if (filter) {
            switch (filter.type) {
                case application.typeFilters.inputText:
                    conditions[filter.nameFilter] = {
                        $regex: values[index],
                        $options: "i",
                    };
                    break;
                case application.typeFilters.select:
                    conditions[filter.nameFilter] = {
                        $eq: values[index],
                    };
                    break;
                case application.typeFilters.selectBool:
                    conditions[filter.nameFilter] = values[index];
                    break;
                case application.typeFilters.dateStart:
                    conditions[filter.nameFilter] = {
                        $gte: values[index],
                        ...conditions[filter.nameFilter],
                    };
                    break;
                case application.typeFilters.dateEnd:
                    conditions[filter.nameFilter] = {
                        $lte: values[index],
                        ...conditions[filter.nameFilter],
                    };
                    break;
                case application.typeFilters.selectMultiple:
                    conditions[filter.nameFilter] = {
                        $in: JSON.parse(values[index]),
                    };
                    break;
                default:
            }
        }
    });

    return conditions;
};

const sortBy = (order, sort) => {
    const orderObject = {};
    orderObject[sort] = order === "desc" ? -1 : 1;
    return orderObject;
};

export default {
    createFilterConditions,
    sortBy,
};
