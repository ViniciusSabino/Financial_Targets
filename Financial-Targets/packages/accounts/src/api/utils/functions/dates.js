import moment from 'moment';

const getCurrentMonthOrADate = (date = moment().format()) => moment(date).month() + 1;

const getCurrentYearOrADate = (date = moment().format()) => moment(date).year();

export { getCurrentMonthOrADate, getCurrentYearOrADate };
