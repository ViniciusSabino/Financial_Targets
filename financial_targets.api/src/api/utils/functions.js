import moment from 'moment';

const getCurrentDate = () => moment().format();

const getDaysInCurrentMonth = () => moment().daysInMonth();

const createMomentDate = date => moment(date);

export default {
  getCurrentDate,
  getDaysInCurrentMonth,
  createMomentDate
};
