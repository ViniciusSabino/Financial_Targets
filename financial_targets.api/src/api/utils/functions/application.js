const result = (data, page = 1, perPage = 20) => ({
    count: data?.length ? data.length : 0,
    page,
    perPage,
    data
});

export default {
    result
};
