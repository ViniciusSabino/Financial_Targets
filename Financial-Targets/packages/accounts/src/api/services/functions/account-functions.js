import { getDaysInCurrentMonth, createMomentDate } from '../../utils/functions/dates';

import { accountEnum, applicationEnum } from '../../utils/enumerators';

const {
    date: { differences },
} = applicationEnum;

const setAccountDate = (dueDate, type) => {
    const days = getDaysInCurrentMonth();

    const dueDateMoment = createMomentDate(dueDate);

    return type === accountEnum.type.monthly
        ? dueDateMoment.add(days, differences.days).format()
        : dueDateMoment.add(1, differences.years).format();
};

export { setAccountDate };
