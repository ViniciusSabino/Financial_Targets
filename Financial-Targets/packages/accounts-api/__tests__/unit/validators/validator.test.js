import moment from "moment";

import context from "../../../src/api/__mocks__/context";
import validator from "../../../src/api/validators/accounts";
import { account, accountEmpty } from "../../../src/api/utils/test/get-account";
import { getCurrentDate } from "../../../src/api/utils/functions/dates";

jest.mock("../../../src/api/utils/functions/dates", () => ({
    getCurrentDate: jest.fn(),
}));

describe("Validator", () => {
    describe("POST -> /account", () => {
        it("should return status 400 and an array of errors when an invalid account is registered", async () => {
            const ctx = {
                ...context,
                request: {
                    body: accountEmpty,
                },
            };

            const next = jest.fn(() => {
                expect(ctx).toMatchSnapshot();
            });

            ctx.badRequest = (payload) => {
                ctx.response.message = payload;
                ctx.response.status = 400;
            };

            await validator.validCreate(ctx, next);

            expect(ctx.response.status).toEqual(400);
            expect(ctx.response.message.errors.length).toEqual(6);
        });

        it("should return an account without a request body", async () => {
            const ctx = {
                ...context,
                request: {
                    body: account,
                },
            };

            ctx.response.status = 200;
            ctx.response.message = "Ok";

            const next = jest.fn(() => {
                expect(ctx).toMatchSnapshot({
                    ...ctx,
                    response: {
                        ...ctx.response,
                        status: 200,
                        message: "Ok",
                    },
                });
            });

            await validator.validCreate(ctx, next);

            expect(ctx.response.status).not.toEqual(400);
        });

        it("should return the done status when the amount paid is equal to the value of the account", async () => {
            const ctx = {
                ...context,
                request: {
                    body: {
                        ...account,
                        amountPaid: 100,
                        value: 100,
                    },
                },
            };

            const next = jest.fn(() => {
                expect(ctx).toMatchSnapshot();
            });

            await validator.validCreate(ctx, next);

            expect(ctx.request.body.status).toEqual("DONE");
        });

        it("should return expired status when the payment date is less than the current date", async () => {
            const ctx = {
                ...context,
                request: {
                    body: {
                        ...account,
                    },
                },
            };

            const next = jest.fn(() => {
                expect(ctx).toMatchSnapshot();
            });

            getCurrentDate.mockImplementation(() =>
                moment(new Date("2019-05-09T03:24:00.000")).format()
            );

            await validator.validCreate(ctx, next);

            expect(ctx.request.body.status).toEqual("EXPIRED");
        });
    });

    describe("GET -> /accounts and /account/all", () => {
        it("should have the user id before listing accounts", async () => {
            const ctx = {
                ...context,
                request: {
                    header: {
                        userid: 1,
                    },
                },
            };

            const next = jest.fn(() => {
                expect(ctx).toMatchSnapshot();
            });

            await validator.validList(ctx, next);

            expect(ctx.request.header.userid).toEqual(1);
        });

        it("should return a bad request otherwise have the id of the user before listing", async () => {
            const ctx = {
                ...context,
                request: {
                    header: {
                        userid: 0,
                    },
                },
            };

            const next = jest.fn(() => {
                expect(ctx).toMatchSnapshot();
            });

            ctx.badRequest = (payload) => {
                ctx.response.message = payload;
                ctx.response.status = 400;
            };

            await validator.validList(ctx, next);

            expect(ctx.response.status).toEqual(400);
            expect(ctx.response.message.errors.length).toEqual(1);
        });
    });

    describe("PUT -> /accounts", () => {
        it("should return status 400 and an array of errors when an invalid account is edit", async () => {
            const ctx = {
                ...context,
                request: {
                    body: accountEmpty,
                },
            };

            const next = jest.fn(() => {
                expect(ctx).toMatchSnapshot();
            });

            ctx.badRequest = (payload) => {
                ctx.response.message = payload;
                ctx.response.status = 400;
            };

            await validator.validEdit(ctx, next);

            expect(ctx.response.status).toEqual(400);
            expect(ctx.response.message.errors.length).toEqual(7);
        });

        it("should return a bad request if the account date is changed to less than current date", async () => {
            const ctx = {
                ...context,
                request: {
                    body: account,
                },
            };

            const next = jest.fn(() => {
                expect(ctx).toMatchSnapshot();
            });

            getCurrentDate.mockImplementation(() =>
                moment(new Date("2019-05-09T03:24:00.000")).format()
            );

            ctx.badRequest = (payload) => {
                ctx.response.message = payload;
                ctx.response.status = 400;
            };

            await validator.validEdit(ctx, next);

            expect(ctx.response.status).toEqual(400);
            expect(ctx.response.message.errors.length).toEqual(1);
        });
    });

    describe("PATCH -> /accounts/makepartialpayment", () => {
        it("should have the acount id before make partial payments", async () => {
            const ctx = {
                ...context,
                request: {
                    body: {
                        accountId: 1,
                    },
                },
            };

            const next = jest.fn(() => {
                expect(ctx).toMatchSnapshot();
            });

            await validator.validMakePartialPayment(ctx, next);

            expect(ctx.request.body.accountId).toEqual(1);
        });

        it("should return a bad request otherwise have the id of the account before make partial payments", async () => {
            const ctx = {
                ...context,
                request: {
                    body: {
                        accountId: 0,
                    },
                },
            };

            const next = jest.fn(() => {
                expect(ctx).toMatchSnapshot();
            });

            ctx.badRequest = (payload) => {
                ctx.response.message = payload;
                ctx.response.status = 400;
            };

            await validator.validMakePartialPayment(ctx, next);

            expect(ctx.response.status).toEqual(400);
            expect(ctx.response.message.errors.length).toEqual(1);
        });
    });
});
