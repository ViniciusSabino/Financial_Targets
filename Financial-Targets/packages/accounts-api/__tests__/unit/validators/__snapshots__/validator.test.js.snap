// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validator GET -> /accounts and /account/all should have the user id before listing accounts 1`] = `
Object {
  "app": Object {
    "env": "development",
    "proxy": false,
    "subdomainOffset": 2,
  },
  "originalUrl": "/accounts",
  "req": "<original node req>",
  "request": Object {
    "header": Object {
      "userid": 1,
    },
  },
  "res": "<original node res>",
  "response": Object {
    "header": Object {},
    "message": "Ok",
    "status": 200,
  },
  "socket": "<original node socket>",
}
`;

exports[`Validator PATCH -> /accounts/makepartialpayment should have the acount id before make partial payments 1`] = `
Object {
  "app": Object {
    "env": "development",
    "proxy": false,
    "subdomainOffset": 2,
  },
  "originalUrl": "/accounts",
  "req": "<original node req>",
  "request": Object {
    "body": Object {
      "accountId": 1,
    },
  },
  "res": "<original node res>",
  "response": Object {
    "header": Object {},
    "message": Object {
      "errors": Array [
        undefined,
      ],
    },
    "status": 400,
  },
  "socket": "<original node socket>",
}
`;

exports[`Validator POST -> /account should return an account without a request body 1`] = `
Object {
  "app": Object {
    "env": "development",
    "proxy": false,
    "subdomainOffset": 2,
  },
  "originalUrl": "/accounts",
  "req": "<original node req>",
  "request": Object {
    "body": Object {
      "amountPaid": 0,
      "description": "Música",
      "dueDate": "2019-05-06T16:54:37-02:00",
      "isRepeat": true,
      "name": "Spotify",
      "paymentForm": "CREDIT",
      "status": "PENDING",
      "tags": Array [
        "Música",
      ],
      "type": "MONTHLY",
      "userId": 1,
      "value": 16.9,
    },
  },
  "res": "<original node res>",
  "response": Object {
    "header": Object {},
    "message": "Ok",
    "status": 200,
  },
  "socket": "<original node socket>",
}
`;

exports[`Validator POST -> /account should return expired status when the payment date is less than the current date 1`] = `
Object {
  "app": Object {
    "env": "development",
    "proxy": false,
    "subdomainOffset": 2,
  },
  "originalUrl": "/accounts",
  "req": "<original node req>",
  "request": Object {
    "body": Object {
      "amountPaid": 0,
      "description": "Música",
      "dueDate": "2019-05-06T16:54:37-02:00",
      "isRepeat": true,
      "name": "Spotify",
      "paymentForm": "CREDIT",
      "status": "EXPIRED",
      "tags": Array [
        "Música",
      ],
      "type": "MONTHLY",
      "userId": 1,
      "value": 16.9,
    },
  },
  "res": "<original node res>",
  "response": Object {
    "header": Object {},
    "message": "Ok",
    "status": 200,
  },
  "socket": "<original node socket>",
}
`;

exports[`Validator POST -> /account should return the done status when the amount paid is equal to the value of the account 1`] = `
Object {
  "app": Object {
    "env": "development",
    "proxy": false,
    "subdomainOffset": 2,
  },
  "originalUrl": "/accounts",
  "req": "<original node req>",
  "request": Object {
    "body": Object {
      "amountPaid": 100,
      "description": "Música",
      "dueDate": "2019-05-06T16:54:37-02:00",
      "isRepeat": true,
      "name": "Spotify",
      "paymentForm": "CREDIT",
      "status": "DONE",
      "tags": Array [
        "Música",
      ],
      "type": "MONTHLY",
      "userId": 1,
      "value": 100,
    },
  },
  "res": "<original node res>",
  "response": Object {
    "header": Object {},
    "message": "Ok",
    "status": 200,
  },
  "socket": "<original node socket>",
}
`;
